### PR TITLE
document mcmc and add some tests

### PR DESCRIPTION
**Pull request by nociale**
@gowerc I documented some functions of the mcmc.R file and added some tests. There are some functions that you implemented to prepare data for the Stan fit. I did not document those as I don't know exactly what the functions do.

### CODE REVIEW

File: R/mcmc.R [hunk 1]
```diff
@@ -114,10 +114,34 @@ fit_mcmc <- function(
...
+#' Transform an array into list of arrays where the listing
+#' is performed on a given dimension.
+#'
+#' @param a Array with number of dimensions at least 2.
+#' @param n Positive integer. Dimension of `a` to be listed.
+#'
+#' @details
+#' For example, if `a` is a 3 dimensional array and `n = 1`,
+#' `split_dim(a,n)` returns a list of 2 dimensional arrays (i.e.
+#' a list of matrices) where each element of the list is `a[i, , ]`, where
```
Comment (gowerc) [inline R/mcmc.R:129]: Its hard to understand what this function is doing from the worded description. Perhaps just provide a small example.  For instance I did something similar in `invert_indexes()`

File: R/mcmc.R [hunk 2]
```diff
@@ -136,29 +160,36 @@ split_dim <- function(a, n) {
 }
 
 
-#' Title - TODO
+#' Extract draws from a `stanfit` object
+#'
+#' @description
+#' Extract draws from a `stanfit` object, manipulate it
```
Comment (gowerc) [inline R/mcmc.R:166]: What manipuilations does it do ?

**Review by gowerc — CHANGES_REQUESTED** [review]

File: R/mcmc.R [hunk 3]
```diff
@@ -136,29 +160,36 @@ split_dim <- function(a, n) {
...
     ##################### from array to list
-    pars$sigma <- split_dim(pars$sigma, 1)
+    pars$sigma <- split_dim(pars$sigma, 1) # list of length equal to the number of draws
 
-    if (length(dim(pars$sigma[[1]])) == 3) { # if same_cov == FALSE
-        pars$sigma <- lapply(
-            pars$sigma,
-            function(x) split_dim(x, 1)
-        )
-    } else {
```
Comment (gowerc) [inline R/mcmc.R:156]: Looks like you've collapsed the logic for same_cov  =TRUe / FALSe into a single block, do you have unit tests that show it work in both scenarios ?

**Review by gowerc — CHANGES_REQUESTED** [review]

File: R/mcmc.R [hunk 4]
```diff
@@ -246,7 +314,11 @@ check_mcmc <- function(stan_fit, n_draws, threshold_lowESS = 0.4) {
 
 #' QR decomposition
 #'
-#' @param mat A Matrix to perform the QR decomposition on
+#' @description
+#' QR decomposition as defined in the
+#' \href{https://mc-stan.org/docs/2_27/stan-users-guide/QR-reparameterization-section.html}{Stan user's guide (section 1.2)}.
```
Comment (gowerc) [inline R/mcmc.R:319]: Minor point, I would recommend using markdown syntax where possible.

```
#' See more about the markdown markup at the
#' [Commonmark web site](http://commonmark.org/help)
```

File: R/mcmc.R (hunk 3 above)
Comment (nociale) [reply R/mcmc.R:156]: Yes, the tests I have added in test-mcmc regards exactly this. Essentially the "else" was never ran since the if condition is always true.

File: R/mcmc.R (hunk 1 above)
Comment (nociale) [reply R/mcmc.R:129]: ok, done

File: R/mcmc.R (hunk 2 above)
Comment (nociale) [reply R/mcmc.R:166]: I re-phrased it better

File: R/mcmc.R [hunk 5]
```diff
@@ -136,29 +187,39 @@ split_dim <- function(a, n) {
 }
 
 
-#' Title - TODO
+#' Extract draws from a `stanfit` object
+#'
+#' @description
+#' Extract draws from a `stanfit` object and convert them into lists.
+#'
+#' The function \link[rstan]{extract} returns the draws for a given parameter as an array. This function
```
Comment (gowerc) [inline R/mcmc.R:195]: Minor but please use the markdown syntax `[rstan::extract()]`

File: R/mcmc.R [hunk 6]
```diff
@@ -136,29 +187,39 @@ split_dim <- function(a, n) {
...
+#' calls \link[rstan]{extract} to extract the draws from a `stanfit` object
```
Comment (gowerc) [inline R/mcmc.R:196]: Please use markdown syntax

File: R/mcmc.R [hunk 7]
```diff
@@ -204,9 +282,20 @@ check_ESS <- function(stan_fit, n_draws, threshold = 0.4) {
...
-#' Title - TODO
+#' Diagnostics of the MCMC based on HMC-related measures.
+#'
+#' @description
+#' Check that:
+#' 1. There are no divergent iterations.
+#' 2. The Bayesian Fraction of Missing Information (BFMI) is sufficiently low.
+#' 3. The number of iterations that saturated the max treedepth is zero.
+#'
+#' Please see \link[rstan]{check_hmc_diagnostics} for details.
```
Comment (gowerc) [inline R/mcmc.R:293]: Please use markdown syntax

File: R/mcmc.R [hunk 8]
```diff
@@ -1,17 +1,52 @@
...
+#' @description
+#' `fit_mcmc()` fits the base imputation model using a Bayesian approach. This is done through a MCMC method that is implemented in `stan`
+#' and is run by using the function [rstan::sampling()].
+#' The function returns the draws from the posterior distribution of the model parameters
+#' and the `stanfit` object. Additionally it performs multiple diagnostics checks of the chain
+#' and returns warnings in case of any detected issues.
+#'
+#' @param designmat The design matrix of the fixed effects.
+#' @param outcome The response variable. Must be numeric.
+#' @param group The group variable.
```
Comment (gowerc) [inline R/mcmc.R:13]: Maybe worth stating what types these need to be,  I.e. I think group and visit need to be factor ?

File: R/mcmc.R [hunk 9]
```diff
@@ -1,17 +1,52 @@
...
+#' 1. Fit MMRM using a REML approach.
+#' 2. Prepare the input data for the MCMC fit as described in the `data{}` block of the Stan file. See [prepare_stan_data()] for details.
+#' 3. Run the MCMC according the input arguments and using as starting values the REML parameter estimates
+#' estimated at point 1.
+#' 4. Performs diagnostics checks of the MCMC. See [check_mcmc()] for details.
+#' 5. Extract the draws from the model fit.
+#'
+#' The chains perform `n_imputations` draws by keeping one every `burn_between` iterations. Additionally
+#' the first `burn_in` iterations are discarded. The total number of iterations will then be `burn_in + burn_between*n_imputations`.
+#' The `burn_in` aims to perform draws from the stationary distribution of the Markov Chain.
```
Comment (gowerc) [inline R/mcmc.R:42]: Maybe reword this slightly?

"the purpose of `burn_in` is to ensure that the samples that are drawn are drawn from the stationary ..." ?

maybe unnecessary though

File: R/mcmc.R [hunk 10]
```diff
@@ -1,17 +1,52 @@
...
+#' The `burn_between` aims to keep the draws uncorrelated each other.
```
Comment (gowerc) [inline R/mcmc.R:43]: missing a "from"

**Comment by gowerc** [comment]
@nociale  I pushed a commit adding documentation for the remaining mcmc functions that I wrote, I hope that it is clear enough ...
